### DISCUSSION TITLE
Dictionaries and Iterators

### DISCUSSION THREAD

**Discussion by mgrant0** [VirusTotal/yara#1720] [Q&A]
I'm trying to extend the yara-json module which parses json using libjansson.  

It currently defines 3 functions:

```
    declare_function("kv", "ss", "i", key_value);  /* return 1 if key==value otherwise 0 */
    declare_function("has_key", "s", "i", has_key);  /* return 1 if key exists */
    declare_function("has_key", "r", "i", has_key_r);  /* return 1 if regex of key exists */
```

Minimal, but a good base to start from.  I'd like to return a dictionary of some subset of the json structure to iterate over in yara.  

The docs don't seem to provide any example on how to do this though it says it's possible (https://yara.readthedocs.io/en/stable/writingrules.html?highlight=iterate#iterators).  I looked at the pe module but it's not obvious to me what's going on.

Does anyone have a concise example how to make a dictionary in a module and how a function returns that or a pointer into that?  It would be great to get an example into the docs for this because it's a great feature!

**Comment by wxsBSD** [comment]
If you're using a reasonable modern version you can use iterators, which are documented here: https://yara.readthedocs.io/en/stable/writingrules.html#iterators

It also talks about iterating over dictionaries. Here's an example of using it to print the `pe.version_info` dictionary:

```
import "console"
import "pe"

rule a {
  condition:
    for all k, v in pe.version_info:
      (console.log("Key: ", k) and console.log("\tValue: ", v))
}
```

**Reply by wxsBSD** [reply]
Sorry, hit enter too soon. To answer the second part of your question I would suggest looking at https://github.com/VirusTotal/yara/blob/master/libyara/modules/pe/pe.c#L3409 for how to define your dictionary and setting the keys and values with https://github.com/VirusTotal/yara/blob/master/libyara/modules/pe/pe.c#L678.

**Reply by mgrant0** [reply]
Ok thanks.  I've read through this now.  I've still got a question/concern here.

Imagine that I have an arbitrary JSON structure.  At the top level it's an object.  Inside the object it could be anything (an array, an object, one or more values..etc).  I am going to iterate over this object in C.  I need to now call functions in the yara api to insert things into a dictionary.  That dictionary being the top level object I think.  For example:

```
{ "foo": "one", "bar": [ 1, 2, 3 ] }
```
The module is named `json` so I am expecting to be able to get at json.foo and iterate over json.bar.  Perhaps being used something like this:

```
json.foo == "one"
for any i in json.bar : ( i == 2)
```

If I understand correctly (which I may not!) it looks like I would define the structure of dictionary using macros defined in module.h.

In my code, module_load() gets the json structure.  How can I then dynamically build the structure which is inside the begin_declarations/end_declarations macros?  I see how to  use set_string(), set_integer()..etc after the structure has been created but I don't see how to create this structure programmatically versus statically.

**Reply by wxsBSD** [reply]
If you know that the value for `foo` will always be a string and the value for `bar` will always be an array of integers you could do something like this:

```
    begin_struct_dictionary("object");
        declare_integer("foo");
        declare_integer_array("bar");
    end_struct_dictionary("object");
```

Then you can populate them by iterating over the JSON in module_load and doing something like (psuedo-code):

```
for key, value in json:
    if key == "foo":
        set_string(value, module, "object[%s]", key)
    if key == "bar":
        for i, value in enumerate(values):
            set_integer(value, module, "object[%s][%i]", key, i)
```

Basically, you just set the value using the key to index into the struct (which I have named object). This is easy in the case of a string, but when adding to the `bar` integer array you need to use the key into the struct and then another one to index into the value array at that key.

This would allow you to express things like `json.object["foo"] == "one"` and `for any i in json.object["bar"]: (i == 2)` in your rules.

This works fine when you know the types of each value in the JSON ahead of time, because you can declare your struct to match accordingly, and iterate over the JSON to translate it into YARA structures.

The problem is if you want this to be a generic JSON parser that can work on any arbitrary JSON structure, where you don't know the types of the values ahead of time. For example, the code above assumes the value at `foo` will be a string, but if it isn't then you're not going to have a good day. ;)

I've been mulling over the idea of a generic JSON parser module for a while now and keep running into the problem described here, where you can't possibly know the structure of the JSON you're parsing ahead of time. In this case I've been thinking about some kind of query interface where you can pass a selector (maybe jq syntax, maybe something else) and it will, on the fly, parse the JSON using that selector and return the results to you. This has problems where we still don't know what the type of the value specified by your selector will be (which we need to know to expose it back via our module function) but maybe there is a way to deal with that too.

Hopefully the first part helps with your question, and highlights some potential problems you might run into if you are trying to make this a generic JSON parser.

**Reply by mgrant0** [reply]
Exactly.  I need to do this for generic JSON.  

I've been looking at the macros in modules.h.  I was wondering, could these macros be somehow inside the module_load() function?  It looks like it stores the declarations inside an array named `stack` but `stack` goes away when the module_declarations() function exits and so I'm a little confused to its purpose.

In order to recurse down into the JSON structure, I would need to call some functions recursively.  It looks like I could walk down the JSON structure and call yr_object_create().  I could either pass stack around or declare it in C at the file level rather than inside a function so it'd be available like a global variable.

Does this stand a chance at working?

**Reply by mgrant0** [reply]
By the way, the yara-json module uses libjansson which works fine.

```
int module_load(YR_SCAN_CONTEXT* context, YR_OBJECT* module_object, void* module_data, size_t module_data_size) {
    YR_MEMORY_BLOCK* block = first_memory_block(context);
    const uint8_t* block_data = block->fetch_data(block);

    json_error_t json_error;
    json_t* json = json_loads((const char*) block_data, 0, &json_error);

    if(!json) {
        fprintf(stderr, "JSON error: on line %d: %s\n", json_error.line, json_error.text);
        return ERROR_INVALID_FORMAT;
    }

    module_object->data = json;

   /* here is where I want to walk through `json` and create the yara object */

    return ERROR_SUCCESS;
}
```

**Reply by wxsBSD** [reply]
> Exactly.  I need to do this for generic JSON.  
> 
> 
> 
> I've been looking at the macros in modules.h.  I was wondering, could these macros be somehow inside the module_load() function?  It looks like it stores the declarations inside an array named `stack` but `stack` goes away when the module_declarations() function exits and so I'm a little confused to its purpose.

The purpose is to declare the interface for your module exposed in the yara language. I honestly don't recall how the declaration macros work, but you can declare all that you want in module load and then set them after you parse the json. This is done heavily in the PE module where everything is declared in module_load() and set in various functions, so you can do it for well defined things...

> In order to recurse down into the JSON structure, I would need to call some functions recursively.  It looks like I could walk down the JSON structure and call yr_object_create().  I could either pass stack around or declare it in C at the file level rather than inside a function so it'd be available like a global variable.

I don't think a global is what you want. But I also think you're missing the interface problem, which needs to be solved before you look into any implementation.

Imagine you have this {"foo": "bar"}. You could represent that with json.object["foo"] and the value would be a string. But what happens when you parse {"foo": 1}? Now you have two different types to store in the same key in your dictionary. What happens when the key is changed to "bar"?

This is just a simple example, but it gets more complex. Imagine parsing this: [{"foo": 1}]. Now you would need an array of dictionaries declared in you module.

This is why I think a generic representation of the JSON structure exposed in yara is not going to work, and was thinking more about generic query functions exposed in a module.

This is also why things like the cuckoo module work only with a very specific json structure. It isn't a generic json parser exposed in yara, but is rather an implementation specific to the way cuckoo reports are structured.

We ended up building something similar to the cuckoo module at work but it was for reports formatted specifically from our sandbox, it wouldn't work with arbitrary json.

I'm not trying to dodge your question, just trying to point out that you should think more about how you want to be able to express arbitrary json in yara, because I think you will find that it isn't going to work as a static set of structures or arrays declared in your module.

**Reply by wxsBSD** [reply]
> By the way, the yara-json module uses libjansson which works fine.
> 
> 
> 
> ```
> 
> int module_load(YR_SCAN_CONTEXT* context, YR_OBJECT* module_object, void* module_data, size_t module_data_size) {
> 
>     YR_MEMORY_BLOCK* block = first_memory_block(context);
> 
>     const uint8_t* block_data = block->fetch_data(block);
> 
> 
> 
>     json_error_t json_error;
> 
>     json_t* json = json_loads((const char*) block_data, 0, &json_error);
> 
> 
> 
>     if(!json) {
> 
>         fprintf(stderr, "JSON error: on line %d: %s\n", json_error.line, json_error.text);
> 
>         return ERROR_INVALID_FORMAT;
> 
>     }
> 
> 
> 
>     module_object->data = json;
> 
> 
> 
>    /* here is where I want to walk through `json` and create the yara object */
> 
> 
> 
>     return ERROR_SUCCESS;
> 
> }
> 
> ```

Yeah, the json parser is not the problem here. The problem is how do you reasonably expose arbitrary json in a set of static structures and arrays declared ahead of time?

**Reply by mgrant0** [reply]
I'm part way through trying to implement this and I hit a possible problem.

The module name is `json`.

If the JSON being read is like `{"foo":"bar"}` it's pretty clear to access foo, it's `json.foo` which is "bar".

But what if the JSON being read is an array or some other element without a name, for example `[1,2,3]`?  How would you access the second element?  `json.[1]`? it's a little weird.  or i wonder if `json[1]` would be possible?

The situation becomes stranger when the JSON is a simple string or number such as `"foo"`.  Would you then access the value "foo" using `json.`?

I'm tempted to put a word in for the first part, for example "val".  If the JSON is an object as in `{"foo":"bar","baz":1}` then the first part would be object and to access "foo" you would use `json.val.foo`.  If the JSON contained just a string `"foo"`, then you would use `json.val` and get "foo".  Best I can think of at the moment.  Any better suggestions?

**Reply by mgrant0** [reply]
I see our replies crossed, I just saw your reply after I posted my reply.

Yes, I totally agree now that I started digging into it.  I see the problem with the names when there aren't names.  I'm wondering if my idea of using "json.val" in front of things can work.  Array of arrays like `[1,2,[3,4]]` get referred to like `json.val.[2][1]`.  Your example `[{"foo": 1}]` would be referred to as `json.val[0].foo`.  The top level object is an array named "val" and inside val, there is an element which is an object that is not named because it's array element 0, then the next level inside the object is an element whose name is "foo" and value 1.

**Reply by wxsBSD** [reply]
But the problem is this can get arbitrarily deep and arbitrarily complex. This is why I keep ending up at a selector language like what jq has for finding the parts of interest, rather than trying to model arbitrary json like you are.

Even if you use a selector function you need to be able to handle arbitrary return values which aren't known until after the selector has found the right values for you.

It's a complex problem for sure, and I don't mean to discourage you from exploring it. Just trying to point out potential other solutions that may result in a better user experience.

For example: {"foo": {"bar": 1}}

For that example I think this is a nice way to express it (and is directly inspired by jq):

json.query("foo.bar") == 1

This works when you know ahead of time that the query function will return an integer, but what happens when you use the same query on

{"foo": {"bar": "baz"}}

Now you need to overload the query function (which you can do) but you won't know which type to return (integer or string) until you actually parse the json in the query function.

I realize I barely answered your question so my apologies for that. Hopefully I've been able to shed some light on this and help you somehow. :)

**Reply by mgrant0** [reply]
You are helping.  Just discussing this here is a huge help.  Let's keep going...

Ok I now totally see your point about the types.  Let's talk about using jq...

There is a not-so-well documented C API for jq: https://github.com/stedolan/jq/wiki/C-API:-jq-program-invocation.  There's also a debian package libjq-dev which contains this lib.  So it should in theory be easy to link in and call from a yara module in place of libjansson.

Yara supports only these return types from a function: string, integer, and float.  But if you want to iterate over something, yara has dictionaries and arrays.  There's also structure types but i'm not sure if they are iterable.  But you can't return dictionaries, arrays or structures from functions.

This kind of gets back to one of my original questions: how do you return something iterable back to yara?  From what I understood, a function can't do this (yet?).  

From what I understand, to iterate over, or do things like `1 of them`, the data has to be stored in something like `json.val.foo.bar` which is not access via a function which I could call something in libjq.  This is where I got pushed down this path of unpacking the JSON into Yara's structures.

Another point against using `json.val.foo.bar` is you can't do regular expressions on the name like you can with a jq query or yara itself.

Do we need to propose some major change for yara 5.x to support this?  Have any better suggestions?

**Reply by wxsBSD** [reply]
> You are helping.  Just discussing this here is a huge help.  Let's keep going...
> 
> 
> 
> Ok I now totally see your point about the types.  Let's talk about using jq...
> 
> 
> 
> There is a not-so-well documented C API for jq: https://github.com/stedolan/jq/wiki/C-API:-jq-program-invocation.  There's also a debian package libjq-dev which contains this lib.  So it should in theory be easy to link in and call from a yara module in place of libjansson.

Yup, this is exactly where I ended up. I wasn't convinced I liked the speed of jq compared to something like simdjson but as a POC I wasn't willing to care enough to change anything.

> Yara supports only these return types from a function: string, integer, and float.  But if you want to iterate over something, yara has dictionaries and arrays.  There's also structure types but i'm not sure if they are iterable.  But you can't return dictionaries, arrays or structures from functions.

Correct, and this was something I also identified. My initial thought was to limit it such that the function(s) only returned scalar types supported by YARA. If your query resulted in a list we would either flatten it to a string (which seems wrong) or return undefined. As a POC I thought that would be fine.

> This kind of gets back to one of my original questions: how do you return something iterable back to yara?  From what I understood, a function can't do this (yet?).  

You identified a critical point I was hoping you would get to. ;)

You can't return an array or struct via a function (at least last time I looked).

> From what I understand, to iterate over, or do things like `1 of them`, the data has to be stored in something like `json.val.foo.bar` which is not access via a function which I could call something in libjq.  This is where I got pushed down this path of unpacking the JSON into Yara's structures.

You are correct here too.

I think if you're willing to eliminate some functionality and only support queries that result in a scalar then you have a better shot at making it work and identifying areas of improvement (like returning arrays via functions). Basically, don't let perfect be the enemy of good while you're exploring here.

> Do we need to propose some major change for yara 5.x to support this?  Have any better suggestions?

It's possible but until you have a demonstrated need for them I wouldn't really pursue them beyond this thread. I think building it and demonstrating the current limitations is better than designing improvements that are theoretical.

**Comment by mgrant0** [comment]
Ok so I have confirmed more of what you said.  I wrote some functions to unpack the json in the module_load() function and here's what I learned:

I learned that the definitions in module_definitions() are defined BEFORE the file is read in module_load().  One idea we talked about above was creating a structure like `json.val.foo.bar`.  This does not work because the file isn't read (the __context isn't even created) until after the yara rules file is parsed.  `json.val.anything` returns

```
error: rule "jsonrule" in test.yara(5): invalid field name "val"
```

Even though I could create val in the module_load() function, it's too late in the game.  As you say, you need to pre-load the structure of the JSON for this to even stand a chance at working.  Like a true catch-22.  I'm abandoning this line of thought here.

It sure would be nice if functions could return something iterable.  But let's see how far I get without that.

> Correct, and this was something I also identified. My initial thought was to limit it such that the function(s) only returned scalar types supported by YARA. If your query resulted in a list we would either flatten it to a string (which seems wrong) or return undefined. As a POC I thought that would be fine.

At this point I have to agree.  However, if the query resulted in something larger than a scalar type, I would return the JSON as a string.  The reason I would do this is at minimum, it allows me to then stick that in a variable and then match a RE on that whole block of JSON.  It's not ideal but it's functional.

libjansson does not support any query into the JSON it loads.  I'm going to look around for a more robust JSON lib.  Even if I find something that returns some structure, I can "flatten" it, or rather cat it together back into a single string and return that as a string (a "sub-JSON" string, a smaller JSON block).  If you have any suggestions of a decent C JSON lib, I'd like to know.

With that, it should make yara quite minimally useful for my purpose (and perhaps others) with JSON formatted data.

**Reply by mgrant0** [reply]
It turns out that I do need to return something from a function that I can iterate over.  For example, if I dig into some JSON and return an array of integers, I now need to do something like find the max of those integers.  Of course I could do that within my module but this defeats the purpose of YARA.  

It's now starting to feel like I need to abandon YARA and find another rules engine.

**Reply by wxsBSD** [reply]
> It sure would be nice if functions could return something iterable. But let's see how far I get without that.

I agree, but probably not worth implementing just yet. ;)
 
> > Correct, and this was something I also identified. My initial thought was to limit it such that the function(s) only returned scalar types supported by YARA. If your query resulted in a list we would either flatten it to a string (which seems wrong) or return undefined. As a POC I thought that would be fine.
> 
> At this point I have to agree. However, if the query resulted in something larger than a scalar type, I would return the JSON as a string. The reason I would do this is at minimum, it allows me to then stick that in a variable and then match a RE on that whole block of JSON. It's not ideal but it's functional.

Yup, not ideal but functional and a starting point. It can always be iterated on as your understanding grows.

> libjansson does not support any query into the JSON it loads. I'm going to look around for a more robust JSON lib. Even if I find something that returns some structure, I can "flatten" it, or rather cat it together back into a single string and return that as a string (a "sub-JSON" string, a smaller JSON block). If you have any suggestions of a decent C JSON lib, I'd like to know.

To be honest, I didn't do any significant research into a robust, fast library with a selector interface. I ended up on `jq` because it is something I have experience with. I have a friend who highly recommends `simdjson` for fast parsing but I have no idea about a selector interface for it.

> With that, it should make yara quite minimally useful for my purpose (and perhaps others) with JSON formatted data.

I'd love to see this work when you're ready to share it!

**Reply by mgrant0** [reply]
I've got some good news.  I've made some excellent forward progress.  I moved to the C library json-s.  json-s has a simple path parser which allows printf/scanf like formats.  I've managed to write a yara function `json.query_d("/foo/bar/%d/baz",0)` when run on:
```
{
  "foo": {
    "bar": [ 
      {
        "baz": "hi"
      },
...
}
```
it returns "hi".  

```
begin_declarations;
    declare_function("query", "s", "s", query_s);
    declare_function("query_s", "ss", "s", query_s);
    declare_function("query_d", "si", "s", query_d);
end_declarations;
```
```
import "json"

rule jsonrule {
        condition:
                json.query("/foo/bar/%d/baz",0) == "hi"
}
```
matches.  This means I should be able to iterate over an array in a for loop.  I don't know if I could iterate over an object's keys/values.

There's no way to have infinite arg formats, I have to create a query_x function and define_function it for the useful permutations (e.g. query_s for a format containing a single %s, query_sd for a format string containing a %s and %d in that order...etc).  If I could have gotten at the argument format (the string containing the "s", "ss", "si"..) so I could call the corresponding x_argument(n) function, I could have done this all in a single function rather than one for each argument arrangement.  Then there's the return value types...

I must admit that using yara for this feels more like using it for something it definitely wasn't designed for (which it wasn't!).

**Reply by wxsBSD** [reply]
Yeah, it feels like a bit of a stretch to have this in yara. One thing that could be useful here is to be able to combine detections from static content and json representation of dynamic execution traces in a single rule.

**Reply by mgrant0** [reply]
I'm not sure what you mean, could you give me an example?

**Reply by mgrant0** [reply]
Do you mean a file which could contain json formatted data as part of a file but not the whole file?  For example something like:

```
; example of json embedded in some other file
foo = { [ { "bar": 1 } ] }
```
Running this through the module I wrote bombs out because the entire file isn't valid JSON. 

What if, instead of reading in the file in module_load(), what if I provided a separate function to load the JSON from a string?  Where would you call such a function?  In the Strings definitions section?  In the Condition section?  Alternatively, and this might be the best solution (but certainly the slowest) would be to pass in the JSON each time you call the query function, for example:

```
import "json"

rule jsonrule {
    strings:
        $j = /\{.*\}/
    condition:
        json.query($j, "/foo/bar/%d/baz",0) == "hi"
}
```
so $j extracts some part of the file which has json, in this contrived example {...}.  Then $j is fed into json.query which would first extract the JSON into a structure in memory and then run json_object_get_string on the in memory representation, then free the in memory representation.  Clearly if you were going to call json.query() many times with the same $j it's wasteful but this could just be the price to pay for this functionality.  I'm not sure how to extract the json into memory and then subsequently call json.query() on it.

**Reply by mgrant0** [reply]
By the way, I tried to call json.query() up in the strings: section with my existing code and it doesn't work.  Apparently you can only call module functions in the condition: section.  Is this your understanding as well?

**Reply by wxsBSD** [reply]
Sorry, been traveling a bit so I missed this.

> Do you mean a file which could contain json formatted data as part of a file but not the whole file?

Nope, that would be a lot more work, to find valid JSON in arbitrary data and parse it. What I meant was how you can use the `-x` argument to wire up a file as module data, which can then be handled via the module. We were using this (at least I think we were) to write rules that combined content of the binary (using normal yara strings) with information from the sandbox report using `$string and sandbox.write_file(/foo.exe/)` conditions.

> By the way, I tried to call json.query() up in the strings: section with my existing code and it doesn't work. Apparently you can only call module functions in the condition: section. Is this your understanding as well?

Yes, that is correct.

**Reply by mgrant0** [reply]
I think I see what cuckoo does to use the -x option.  It stores the JSON in an already defined structure.  From the point of view, it's the same as when loading in the actual file being read.  You would need to have already defined that structure before reading module_load() was called.  It's the same exact problem with reading in the file itself.

In my experimentation, the -x json:foo.json, what this does is pass foo.json in the module_data param of the module_load() function call.  

As far as I can tell, the -x doesn't help.  Have I missed something?

Do you know of any other languages like Yara that I might want to look at?  Something more specific to JSON perhaps?

**Reply by wxsBSD** [reply]
> I think I see what cuckoo does to use the -x option.  It stores the JSON in an already defined structure.  From the point of view, it's the same as when loading in the actual file being read.  You would need to have already defined that structure before reading module_load() was called.  It's the same exact problem with reading in the file itself.

Yes, you still have the problems associated with not knowing the structure of the json ahead of time. My point in suggesting why a generic json parser might be useful is you can use the module data trick to pass json to that module AND pass arbitrary data to the rest of yara. This way you can express things like "$string and json.foo()" where the string is searched for in the arbitrary data and the json can be something related to it (like say a sandbox report).

This won't solve your problems as you've found out but it does combine static detections from yara with arbitrary json expressions over the output of some other tool. For example, being able to express something like "pe.imphash() == "foo" and json.writes_file(/bar/)" is an interesting capability.

> In my experimentation, the -x json:foo.json, what this does is pass foo.json in the module_data param of the module_load() function call.  

Correct. But for the reason I highlighted above it is probably the best option here.

> As far as I can tell, the -x doesn't help.  Have I missed something?

It doesn't address the problem of arbitrary json structures, but it does open up possibilities if you can live with the limitations already identified.

> Do you know of any other languages like Yara that I might want to look at?  Something more specific to JSON perhaps?

jq? Maybe something that kicks off a yara scan and executed arbitrary jq expressions in parallel and then combines the results?
